xtask/src/release.rs

```diff
@@ -1,63 +1,60 @@
-        println!("Making GitHub release for {} {}…", self.name, version);
-
-        let changes = self.get_changes(&version)?;
-
-        let tag = format!("{}-{}", self.name, version);
-        let name = format!("{} {}", self.name, version);
-
-        cmd!("git tag").args(&["-a", &tag, "-m", &name, "-m", &changes]).secret(true).run()?;
-
-        cmd!("git push").args(&[&config.remote, &tag]).run()?;
-
-        cmd!("curl")
-            .args(&[
-                "-u",
-                &format!("{}:{}", config.user, config.token),
-                "-X",
-                "POST",
-                "-H",
-                "Accept: application/vnd.github.v3+json",
-                &format!("https://api.github.com/repos/{}/releases", config.repo),
-                "-d",
-                &json!({
-                    "tag_name": tag,
-                    "name": name,
-                    "body": changes,
-                })
-                .to_string(),
-            ])
-            .secret(true)
-            .run()?;
+        let _dir = pushd(project_root()?.join(&self.name))?;
 
-        Ok(())
-    }
+        let remote = &self.get_remote()?;
 
-    /// Get the current version of the crate from the manifest.
-    fn get_version(&self) -> Result<String> {
-        let manifest = read_file(project_root().join(&self.name).join("Cargo.toml"))?;
-        for line in manifest.lines() {
-            if line.starts_with("version = ") {
-                return Ok(line[10..].trim_matches('"').into());
-            }
+        if !cmd!("git status -s -uno").read()?.is_empty() {
+            Err("This git repository contains untracked files")?
         }
 
-        Err("Could not find crate version in manifest")?
+        let version = &self.get_version()?;
+        println!("Making release for {} {}…", self.name, version);
+
+        cmd!("cargo publish").run()?;
+
+        let credentials = &config()?.github.credentials();
+
+        let changes = &self.get_changes(&version)?;
+
+        let tag = &format!("{}-{}", self.name, version);
+        let name = &format!("{} {}", self.name, version);
+
+        cmd!("git tag -s {tag} -m {name} -m {changes}").secret(true).run()?;
+
+        cmd!("git push {remote} {tag}").run()?;
+
+        let request_body = &json!({
+            "tag_name": tag,
+            "name": name,
+            "body": changes.trim_softbreaks(),
+        })
+        .to_string();
+
+        cmd!(
+            "curl -u {credentials} -X POST -H 'Accept: application/vnd.github.v3+json'
+            {GITHUB_API_RELEASES} -d {request_body}"
+        )
+        .secret(true)
+        .run()?;
+
+        Ok(())
     }
 
     /// Get the changes of the given version from the changelog.
     fn get_changes(&self, version: &str) -> Result<String> {
-        let changelog = read_file(project_root().join(&self.name).join("CHANGELOG.md"))?;
+        let changelog = read_file(project_root()?.join(&self.name).join("CHANGELOG.md"))?;
```

Comment:
I don't like `project_root` being called multiple times. It should be called once at the start of the release command, on in `main`, and then the value should be passed around.